src/catchup/simulation/ApplyTransactionsWork.cpp

```diff
@@ -35,17 +49,190 @@ ApplyTransactionsWork::ApplyTransactionsWork(
     }
 }
 
+static void
+checkResults(Application& app, uint32_t ledger,
+             std::vector<TransactionResultPair> const& results)
+{
+    auto resSet = TransactionFrame::getTransactionHistoryResults(
+        app.getDatabase(), ledger);
+
+    assert(resSet.results.size() == results.size());
+    for (size_t i = 0; i < results.size(); i++)
+    {
+        auto const& res = results[i];
+        assert(res.transactionHash == resSet.results[i].transactionHash);
+
+        auto const& dbRes = resSet.results[i].result;
+        if (dbRes.result.code() != res.result.result.code())
+        {
+            auto msg = fmt::format(
+                "Expected result code {} does not agree with {} for tx {}",
+                res.result.result.code(), dbRes.result.code(),
+                binToHex(res.transactionHash));
+            CLOG(ERROR, "History") << msg;
+        }
+    }
+}
+
+bool
+ApplyTransactionsWork::hasSig(PublicKey const& account,
+                              TransactionEnvelope const& oldEnvelope,
+                              Hash const& hash)
+{
+    // Is the signature of this account present in the envelope we're
+    // simulating?
+    if (!mUsedPubKeys.insert(account).second)
+    {
+        return false;
+    }
+
+    for (auto const& sig : oldEnvelope.signatures)
+    {
+        if (PubKeyUtils::verifySig(account, sig.signature, hash))
+        {
+            return true;
+        }
+    }
+    return false;
+}
+
+void
+ApplyTransactionsWork::addSignerKeys(AccountID const& acc,
+                                     AbstractLedgerTxn& ltx,
+                                     std::vector<SecretKey>& keys,
+                                     TransactionEnvelope const& oldEnvelope,
+                                     uint32_t n)
+{
+    auto opKey = SimulationUtils::getNewSecret(acc, n);
+    auto const& txHash = mResultIter->transactionHash;
+
+    if (hasSig(acc, oldEnvelope, txHash))
+    {
+        keys.emplace_back(opKey);
+    }
+
+    auto account = stellar::loadAccount(ltx, acc);
+    if (!account)
+    {
+        return;
+    }
+
+    for (auto const& signer : account.current().data.account().signers)
+    {
+        if (signer.key.type() == SIGNER_KEY_TYPE_ED25519)
+        {
+            auto pubKey = KeyUtils::convertKey<PublicKey>(signer.key);
+            if (hasSig(pubKey, oldEnvelope, txHash))
+            {
+                keys.emplace_back(SimulationUtils::getNewSecret(pubKey, n));
+            }
+        }
+    }
+}
+
+uint32_t
+ApplyTransactionsWork::scaleLedger(
```

Comment:
Maybe the notes on TxSimulation should be in their own `TxSimulation.md` file somewhere (that we can point to from the perf docs) instead of scattered in the code base